gobblin-core/src/main/java/gobblin/source/extractor/extract/jdbc/TeradataExtractor.java

```diff
@@ -0,0 +1,334 @@
+  private static final String TERADATA_HOUR_FORMAT = "HH";
+  private static final long SAMPLE_RECORD_COUNT = -1;
+  private static final String ELEMENT_DATA_TYPE = "string";
+  
+  private static final String TERADATA_SAMPLE_CLAUSE =" sample "; 
+  
+  private static final Gson gson = new Gson();
+  
+  public TeradataExtractor(WorkUnitState workUnitState) {
+    super(workUnitState);
+  }
+
+  @Override
+  public List<Command> getSchemaMetadata(String schema, String entity) throws SchemaException {
+    log.debug("Build query to get schema");
+    
+    List<Command> commands = new ArrayList<Command>();
+    
+    String inputQuery = this.workUnit.getProp(ConfigurationKeys.SOURCE_QUERYBASED_QUERY);
+    String metadataSql, predicate = "1=0"; 
+    if(isNullOrEmpty(inputQuery)) {
+      metadataSql = "select * from " + schema + "." + entity;
+    } else {
+      metadataSql = this.removeSampleClauseFromQuery(inputQuery);
+    }
+
+    metadataSql = SqlQueryUtils.addPredicate(metadataSql, predicate);
+    commands.add(JdbcExtractor.getCommand(metadataSql, JdbcCommand.JdbcCommandType.QUERY));
+    return commands;
+  }
+
+  @Override
+  public List<Command> getHighWatermarkMetadata(String schema, String entity, String watermarkColumn,
+      List<Predicate> predicateList) throws HighWatermarkException {
+    log.debug("Build query to get high watermark");
+    List<Command> commands = new ArrayList<Command>();
+
+    String columnProjection = "max(" + Utils.getCoalesceColumnNames(watermarkColumn) + ")";
+    String watermarkFilter = this.concatPredicates(predicateList);
+    String query = this.getExtractSql();
+
+    if (isNullOrEmpty(watermarkFilter)) {
+      watermarkFilter = "1=1";
+    }
+    query = query.replace(this.getOutputColumnProjection(), columnProjection)
+        .replace(ConfigurationKeys.DEFAULT_SOURCE_QUERYBASED_WATERMARK_PREDICATE_SYMBOL, watermarkFilter);
+
+    commands.add(JdbcExtractor.getCommand(query, JdbcCommand.JdbcCommandType.QUERY));
+    return commands;
+  }
+
+  @Override
+  public List<Command> getCountMetadata(String schema, String entity, WorkUnit workUnit, List<Predicate> predicateList)
+      throws RecordCountException {
+    log.debug("Build query to get source record count");
+    List<Command> commands = new ArrayList<Command>();
+
+    String columnProjection = "COUNT(1)";
+    String watermarkFilter = this.concatPredicates(predicateList);
+    String query = this.getExtractSql();
+
+    if (isNullOrEmpty(watermarkFilter)) {
+      watermarkFilter = "1=1";
+    }
+    query = query.replace(this.getOutputColumnProjection(), columnProjection)
+        .replace(ConfigurationKeys.DEFAULT_SOURCE_QUERYBASED_WATERMARK_PREDICATE_SYMBOL, watermarkFilter);
+    String sampleFilter = this.constructSampleClause();
+    query = query + sampleFilter;
+
+    if (!isNullOrEmpty(sampleFilter)) {
+      query = "SELECT COUNT(1) FROM (" + query.replace(" COUNT(1) ", " 1 at t") + ") temp";
```

Comment:
Teradata count returns an `INT`. For big tables (rows > 2^31-1) this query produces a numerical overflow.
I'd suggest to have `CAST(COUNT(1) AS BIGINT)` instead.